src/libos/crates/sworn-disk/src/index/compaction.rs

```diff
@@ -0,0 +1,214 @@
+//! Lsm tree compaction policy.
+use super::bit_builder::BITBuilder;
+use super::mem_table::LockedMemTable;
+use super::metadata::{BITMeta, SearchFlag};
+use crate::prelude::*;
+use crate::{CacheState, Checkpoint, Record};
+
+use std::collections::{BTreeMap, HashSet};
+
+/// Compactor. Used to execute different compaction policy.
+pub struct Compactor(AsyncMutex<()>);
+
+impl Compactor {
+    pub fn new() -> Self {
+        Self(AsyncMutex::new(()))
+    }
+
+    /// Minor compaction. (immutable MemTable -> L0)
+    pub async fn exec_minor_compaction(
+        &self,
+        immut_mem_table: Arc<LockedMemTable>,
+        disk: Arc<dyn BlockDevice>,
+        checkpoint: Arc<Checkpoint>,
+    ) -> Result<()> {
+        let guard = self.0.lock().await;
+
+        assert!(
+            immut_mem_table.state() == CacheState::Full,
+            "Wrong state[{:?}] of immutable memtable",
+            immut_mem_table.state()
+        );
+        // Update state and clear events
+        immut_mem_table.set_state(CacheState::Flushing);
+        immut_mem_table.clear_events();
+
+        let mut records = immut_mem_table.get_all_records();
+        assert!(
+            records.len() <= MAX_MEM_TABLE_CAPACITY,
+            "Compaction error, wrong memtable size [{}]",
+            records.len()
+        );
+        debug_assert!(records.is_sorted_by_key(|record| record.lba));
+
+        let l0_bit_meta = checkpoint.bitc().write().get_l0_bit_meta();
+
+        // First check if major compaction is needed
+        if l0_bit_meta.is_some() {
+            self.exec_major_compaction(
+                l0_bit_meta.as_ref().unwrap().clone(),
+                disk.clone(),
+                checkpoint.clone(),
+            )
+            .await?;
+        }
+
+        // Pick an available segment from `index SVT`
+        let seg_addr = checkpoint.index_svt().write().pick_avail_seg().unwrap();
```

Comment:
Under current implementation, the size of an on-disk BIT is settled (`INDEX_SEGMENT_SIZE`), it always includes a fixed-size records (since I use fixed-size array rather vector inside BIT nodes) and always manages same amount data (512MB in this case, can be configured).

So the max number of BITs in L1 is calculated at first (always manage all data region), so is index_svt's capacity, BIT start addresses and index region.

So I would need a padding logic in compaction and an exact counter in memtable. And a `BITBuf` has known-size content and can always write to disk successfully.